mlflow/gateway/gateway_app.py

```diff
@@ -0,0 +1,106 @@
+import argparse
+from fastapi import FastAPI, Request, HTTPException
+import os
+from pydantic import BaseModel
+from traceback import format_exc
+from typing import List, Optional
+import uvicorn
+
+from mlflow.gateway.constants import GATEWAY_VERSION, CONF_PATH_ENV_VAR
+from mlflow.gateway.handlers import Route, RouteConfig, _route_config_to_route, _load_gateway_config
+from mlflow.gateway.utils import _parse_url_path_for_base_url
+
+# Configured and initialized Gateway Routes state index
+ACTIVE_ROUTES: List[Route] = []
+
+app = FastAPI(
+    title="MLflow Model Gateway API",
+    description="The core gateway API for reverse proxy interface using remote inference "
+    "endpoints within MLflow",
+    version=GATEWAY_VERSION,
+)
+
+
+@app.get("/health")
+async def health():
+    return {"status": "OK"}
+
+
+@app.get("/gateway_url")
+async def gateway_url(request: Request):
+    return {"url": _parse_url_path_for_base_url(str(request.url))}
+
+
+@app.get("/get_route/{route:path}")
+async def get_route(route: str):
+    filtered = [x for x in ACTIVE_ROUTES if x.name == route]
+    if not filtered:
+        raise HTTPException(
+            status_code=404,
+            detail=f"The route '{route}' is not present or active on the server. Please "
+            "verify the route name.",
+        )
+    return {"route": filtered[0]}
+
+
+@app.get("/search_routes")
+async def search_routes(filter: str):
+    # placeholder route listing functionality
+    routes = [route.json() for route in ACTIVE_ROUTES]
+    return {"routes": routes}
+
+
+@app.get("/list_all_routes")
+async def list_all_routes():
+    routes = []
+    for route in app.routes:
+        routes.append(
+            {
+                "path": route.path,
+                "name": route.name,
+                "methods": sorted(route.methods),
+            }
+        )
+    return routes
+
+
+# Placeholder logic for dynamic route configuration
+def _add_dynamic_route(route: RouteConfig):
+    # TODO: Provide a mapping that returns the constructed async endpoint config from the
+    # provider within the RouteConfig. The `route_endpoint` configuration below is a
+    # placeholder only and should be generated based on the expected endpoint configuration
+    # within the given RouteConfig. The pydantic class below is also a placeholder and should
+    # be externalized.
+
+    class RouteInput(BaseModel):
+        input: str
+        temperature: Optional[float]
+
+    async def route_endpoint(data: RouteInput):
+        try:
+            # TODO: handle model-specific route logic by mapping the provider to plugin logic
+
+            return {"response": data.input}
```

Comment:
@dbczumar @BenWilson2 Let's say I create a route for openai chat (https://api.openai.com/v1/chat/completions). What should the response of this route look like?

Reply:
Can we just return what `https://api.openai.com/v1/chat/completions` returns?

Reply:
We want a provider-agnostic return so that if a project COULD use openai or anthropic, switching between them doesn't mean refactoring their consumer code.